.ci/openshift-ci/buildall_install.sh

```diff
@@ -0,0 +1,76 @@
+#!/bin/bash
+#
+# Copyright (c) 2020 Red Hat, Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# Build Kata Containers and install into a given destination directory.
+#
+
+set -e
+
+export PATH="${GOPATH}/bin:$PATH"
+
+cidir="$(dirname "$0")/.."
+source /etc/os-release || source /usr/lib/os-release
+source "${cidir}/lib.sh"
+
+[ -z "$1" ] && die "Usage: $0 path/to/install/dir"
+export DESTDIR="$1"
+info "Build and install Kata Containers at ${DESTDIR}"
+
+[ "$ID" != "centos" ] && die "Expect the build root to be CentOS"
+# The scripts rely on sudo which is not installed in the build environment.
+yum install -y sudo
+"${cidir}/setup_env_centos.sh" default
+
+# The build root container has already golang installed, let's remove it
+# so that it will use the version required by kata.
+yum remove -y golang
+"${cidir}/install_go.sh" -p -f
+
+# Let's ensure scripts don't try things with Podman.
+export TEST_CGROUPSV2="false"
+
+# Configure to use the standard QEMU.
+export experimental_qemu="false"
+export experimental_kernel="false"
+
+# Configure to use the initrd rootfs.
+export TEST_INITRD="yes"
+
+# Configure to use vsock.
+# TODO: install_runtime.sh will try to load the vsock module and the script
+# fail. See https://github.com/kata-containers/tests/issues/2614
+#export USE_VSOCK="yes"
+
+# Configure the QEMU machine type.
+export MACHINETYPE="q35"
+
+# Disable SELinux.
+export FEATURE_SELINUX="no"
+
+# The default /usr prefix makes the deployment on Red Hat CoreOS (rhcos) more
+# complex because that directory is read-only by design. Another prefix than
+# /opt/kata is problematic too because QEMU experimental eventually got from
+# Kata Containers Jenkins is uncompressed in that directory.
```

Comment:
I don't understand this sentence. Even grammatically.

Do you mean that the CI also installs qemu, and that we need to add our stuff in the same spot (if so, why?). Is it because the prefix will be used for the configuration files, and passing `/opt/kata` "magically" generates configuration files that point to the right qemu? If so, it really needs to be explained. That's what I understood reading the sentence, but clearly. the `sed` below adjusts the configuration to point back to `/usr/bin`. so I'm a bit confused.

From a CI point of view, does that mean that what we test with the CI is not installed in the path where it would be installed on an RHCOS system? What are the implications, if any?

Reply:
> I don't understand this sentence. Even grammatically.

Reading it again, I even didn't understand what I wrote. :)
I will rephrase it.

> 
> Do you mean that the CI also installs qemu, and that we need to add our stuff in the same spot (if so, why?). Is it because the prefix will be used for the configuration files, and passing `/opt/kata` "magically" generates configuration files that point to the right qemu? If so, it really needs to be explained. That's what I understood reading the sentence, but clearly. the `sed` below adjusts the configuration to point back to `/usr/bin`. so I'm a bit confused.
>

The outcome of `buildall_install.sh` is a full installation of Kata Containers under `runtime/_out/build_install`. This is later packaged in a container image along with a install script. By the use of Kubernetes DaemonSets, the content of `runtime/_out/build_install` is rsync'ed with the node's filesystem.

That said, some facts:
 1. By default Kata Containers build to `/usr/` and `PREFIX=/usr`
 2. On the cluster node (rhcos):
   * `/usr` is read-only and should not be changed. (see [rhcos architecture](https://docs.openshift.com/container-platform/4.5/architecture/architecture-rhcos.html))
   * `/opt` is a symlink to `/var/opt`
   * `/var` is writable  but, for the sake of support, it should be changed only by machineconfig objects.
 3. Kata CI jobs use statically built QEMU rather than of the distros. The CI script will simply fetch and uncompress the tarball in the test machine.
 * Kata provides two flavours of QEMU for CI:
   * "experimental" QEMU and virtiofs (see [qemu-experimental-nightly-x86_64](http://jenkins.katacontainers.io/job/qemu-experimental-nightly-x86_64) job) built for `/opt/kata` prefix.
   * "standard" QEMU (see the  [qemu-nightly-x86_64](http://jenkins.katacontainers.io/job/qemu-nightly-x86_64) job) which unlike the experimental QEMU is built for `/usr/`
 4. Kata CI jobs also use built kernel (guest OS) and rootfs (initrd or ramdisk)
   * Alike QEMU those files are fetched from Jenkins and uncompressed in the system. Moreover, they will be uncompressed following the `PREFIX` value.
 5. Based on the value of `PREFIX` the CI scripts will configure Kata's `configuration.toml` so that it contain the right paths to QEMU (virtiofsd, if the case), Kernel and rootfs.

In `buildall_install.sh` script `PREFIX` is set to `/opt/kata` so that Kata is built to `/opt/kata` and then I avoid the `fact 2` (`/usr` being readonly).  The kernel and rootfs are properly copied to under `/opt/kata` (`fact 4`), the "standard" QEMU goes to `/usr` (`fact 3`) -- here is the problem -- and then the `configuration.toml` file gets mis-configured (`fact 5`). So I change the `configuration.toml` to fix that as you can see below.

You might ask:

 Q: How do you install QEMU in `/usr` given it is readonly?
 A: Dirty hack. Notice that on `runtime/.ci/openshift-ci/images/entrypoint.sh` script the `/usr` is re-mounted as read-write

 Q: QEMU is static, so can't you simply copy the binary to `/opt/kata/bin`?
 A: It is possible but there is another issue. It is built the firmwares as well, and the QEMU binary still expect them under `/usr`. Perhaps that can be fixed by changing to firmware location on `configuration.toml` too.

 Q: Can't it be solved differently then?
 A: Yes, if we have a "standard" QEMU built with `/opt/kata` prefix and cached in Jenkins the problem will likely to be over. Honestly, I plan to work on it after this PR gets merged.

 Q: Why don't you use QEMU from rhcos?
 A: It is trick and brings in a completely different string of issues. We can discuss that in details later but, briefly speaking, rhcos doesn't package QEMU. We could use, say, the CentOS package but on openshift 4.5 it doesn't install smoothly (conflicts with glusterfs....).

All in all, there are plenty of improvements and important decisions to be taken but I am deferring them to after I get this merged. Because I think what I propose here works pretty much well as the v0 to get Kata on-boarded on Openshift CI.

Reply:
> * Kata provides two flavours of QEMU for CI:
>   
>   * "experimental" QEMU and virtiofs (see [qemu-experimental-nightly-x86_64](http://jenkins.katacontainers.io/job/qemu-experimental-nightly-x86_64) job) built for `/opt/kata` prefix.
>   * "standard" QEMU (see the  [qemu-nightly-x86_64](http://jenkins.katacontainers.io/job/qemu-nightly-x86_64) job) which unlike the experimental QEMU is built for `/usr/`

That's good news. I thought it was only testing the experimental one.

> In `buildall_install.sh` script `PREFIX` is set to `/opt/kata` so that Kata is built to `/opt/kata` and then I avoid the `fact 2` (`/usr` being readonly). The kernel and rootfs are properly copied to under `/opt/kata` (`fact 4`), the "standard" QEMU goes to `/usr` (`fact 3`) -- here is the problem -- and then the `configuration.toml` file gets mis-configured (`fact 5`). So I change the `configuration.toml` to fix that as you can see below.

A meta Q about that. If RHCOS is supposed to be a read-only system, why does it let you install stuff in /opt at all? :smiley:

> Q: How do you install QEMU in `/usr` given it is readonly?
> A: Dirty hack. Notice that on `runtime/.ci/openshift-ci/images/entrypoint.sh` script the `/usr` is re-mounted as read-write

That part I was aware of, but it's definitely useful to mention it in the comments (it's not obvious at all IMO to someone not familiar with RHCOS)

> 
> Q: QEMU is static, so can't you simply copy the binary to `/opt/kata/bin`?
> A: It is possible but there is another issue. It is built the firmwares as well, and the QEMU binary still expect them under `/usr`. Perhaps that can be fixed by changing to firmware location on `configuration.toml` too.

A longer term solution might be to mount some subset of the container root fs into the target location (to avoid copies while preserving the right paths)


> Q: Can't it be solved differently then?
> A: Yes, if we have a "standard" QEMU built with `/opt/kata` prefix and cached in Jenkins the problem will likely to be over. Honestly, I plan to work on it after this PR gets merged.

Sounds good.

> 
> Q: Why don't you use QEMU from rhcos?
> A: It is trick and brings in a completely different string of issues. We can discuss that in details later but, briefly speaking, rhcos doesn't package QEMU. We could use, say, the CentOS package but on openshift 4.5 it doesn't install smoothly (conflicts with glusterfs....).

That's another Q I did not have personally, but that would be a valuable addition to the commit message.

> 
> All in all, there are plenty of improvements and important decisions to be taken but I am deferring them to after I get this merged. Because I think what I propose here works pretty much well as the v0 to get Kata on-boarded on Openshift CI.

I think that your Q / A is an excellent addition to the commit msg. Thanks a bunch.